String.format now uses dependency graph

allredj opened the pull request:
Dependency graph implementation of String.format.

This is mostly profitable for TG, as it allows to lazy-evaluate `String.format` operations that are not part of the control flow decision.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier commented on src/solvers/strings/string_format_builtin_function.h:
```diff
@@ -0,0 +1,80 @@
+
+  bool maybe_testing_function() const override
+  {
+    return false;
```
This implies that the length of the result is only affected by the length of the inputs and not there content, which is not true for primitive types, this can cause dependencies to not be propagated to the primitive values and we need to rely on the user of this builtin function to explicitly constrain the serialized strings for correctness. This should be document here (and maybe in the module documentation)

allredj replied:
Added documentation. I will also add something to the module documentation.

romainbrenguier commented on unit/solvers/strings/string_format_builtin_function/length_for_format_specifier.cpp:
```diff
@@ -0,0 +1,261 @@
+    string_array.begin(),
+    string_array.end(),
+    std::back_inserter(expr_array),
+    [&](const mp_integer el) { return from_integer(el, char_type); });
```
:pick: could be simpler with `const auto expr_array = make_range(string_array).map([&](const mp_integer el) { return from_integer(el, char_type); }).collect<std::vector<exprt>>()`

allredj replied:
I'll skip that one.

romainbrenguier commented on unit/solvers/strings/string_format_builtin_function/length_for_format_specifier.cpp:
```diff
@@ -0,0 +1,261 @@
+      THEN("length constraint represents 11")
+      {
+        const int to_int = *numeric_cast<int>(to_constant_expr(simpl_actual));
+        REQUIRE(to_int == 11);
```
:pick: I would add a first repquire which check that the `numeric_cast` had a non-empty return

allredj replied:
I don't think that's very likely to happen. I'll keep it like this to keep the unit test simple. It will fail anyway if the conversion fails (although not as gracefully).

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: da42ff91c47889df52329100a3fd4b2349da6133).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/114765663

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: f998beb6cf3ede5bff6aceff25e4a7da6fb250c0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/114779221
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

romainbrenguier commented on src/solvers/strings/string_format_builtin_function.cpp:
```diff
@@ -0,0 +1,505 @@
+                 "arguments of format should be strings");
+
+               auto ex = get_string_expr(array_pool, arg);
+               return ex;
```
the intermediary variable ex is  not useful

romainbrenguier commented on src/solvers/strings/string_format_builtin_function.cpp:
```diff
@@ -0,0 +1,505 @@
+      "format must be bounded by 0xFFFF");
+  }
+
+  int64_t int64_value =
```
const

romainbrenguier commented on src/solvers/strings/string_format_builtin_function.cpp:
```diff
@@ -0,0 +1,505 @@
+  int64_t int64_value =
+    (serialized[0] << 48).to_long() | (serialized[1] << 32).to_long() |
+    (serialized[2] << 16).to_long() | serialized[3].to_long();
+  mp_integer mp_integer_value = (mp_integer)int64_value;
```
`const mp_integer mp_interger_value{int64_value}`? casting the int64_value to an mp_integer seems risky

romainbrenguier commented on src/solvers/strings/string_format_builtin_function.cpp:
```diff
@@ -0,0 +1,505 @@
+    (serialized[0] << 48).to_long() | (serialized[1] << 32).to_long() |
+    (serialized[2] << 16).to_long() | serialized[3].to_long();
+  mp_integer mp_integer_value = (mp_integer)int64_value;
+  std::string long_as_string = integer2string(mp_integer_value, base);
```
const

romainbrenguier commented on src/solvers/strings/string_format_builtin_function.cpp:
```diff
@@ -0,0 +1,505 @@
+      it != long_as_string.rend();
+      it++)
+  {
+    deserialized.insert(deserialized.begin(), (mp_integer)*it);
```
Why go backwark and insert at the begining instead of going forward and inserting at the back?

romainbrenguier commented on src/solvers/strings/string_format_builtin_function.cpp:
```diff
@@ -0,0 +1,505 @@
+  return deserialized;
+}
+
+bool eval_is_null(const std::vector<mp_integer> &string)
```
static

romainbrenguier commented on src/solvers/strings/string_format_builtin_function.cpp:
```diff
@@ -0,0 +1,505 @@
+  case format_specifiert::DATE_TIME:
+    /// \todo Conversion of date-time is not implemented
+    // For all these unimplemented cases we return a non-deterministic string
+    UNIMPLEMENTED;
```
this disagree with the comment above

romainbrenguier commented on src/solvers/strings/string_format_builtin_function.cpp:
```diff
@@ -0,0 +1,505 @@
+          // first argument `args[0]` corresponds to index 1
+          if(auto arg_value = eval_string(inputs[fs.index - 1], get_value))
+          {
+            // TODO call eval_format_specifier
```
shouldn't this be done in the same commit?

romainbrenguier commented on src/solvers/strings/string_format_builtin_function.cpp:
```diff
@@ -0,0 +1,505 @@
+    simplify_expr(result_constraint_pair.first, generator.ns).is_zero(),
+    "add_axioms_for_format should return 0, meaning that formatting was"
+    "successful");
+  generator.constraints.existential.push_back(
```
it would be better to add the constraint in result_constraint_pair.second so that all the generated constraints are in the return value of `constraints` and to minimize side effects

romainbrenguier commented on src/solvers/strings/string_format_builtin_function.cpp:
```diff
@@ -0,0 +1,505 @@
+    pow((mp_integer)radix, max_length - 1);
+  long max_value = max_value_for_smaller_size.to_long();
+  max_value--;
+  max_value++;
```
why?

romainbrenguier commented on src/solvers/strings/string_format_builtin_function.cpp:
```diff
@@ -0,0 +1,505 @@
+  // If the current value doesn't fit in a smaller size representation, we have
+  // found the number of digits needed to represent that value.
+  mp_integer max_value_for_smaller_size =
+    pow((mp_integer)radix, max_length - 1);
```
const

romainbrenguier commented on src/solvers/strings/string_format_builtin_function.cpp:
```diff
@@ -0,0 +1,505 @@
+}
+
+/// Return an new expression representing the length of the representation of
+/// `integer`.
```
there is redundancy with `\return`

romainbrenguier commented on src/solvers/strings/string_format_builtin_function.cpp:
```diff
@@ -0,0 +1,505 @@
+}
+
+/// Return an expression representing the length of the format specifier
+/// Does not assume that arg is constant
```
dot at the end of the sentence

romainbrenguier commented on src/solvers/strings/string_format_builtin_function.cpp:
```diff
@@ -0,0 +1,505 @@
+  {
+    exprt retval = length_of_decimal_int(
+      format_arg_from_string(arg, ID_int, array_pool), index_type, 16);
+    return retval;
```
intermediary retval unecessary

romainbrenguier commented on src/solvers/strings/string_format_builtin_function.cpp:
```diff
@@ -0,0 +1,505 @@
+  case format_specifiert::CHARACTER_UPPER:
+  {
+    exprt arg_string = format_arg_from_string(arg, ID_char, array_pool);
+    array_string_exprt &string_expr = to_array_string_expr(arg_string);
```
both const

romainbrenguier commented on src/solvers/strings/string_format_builtin_function.cpp:
```diff
@@ -0,0 +1,505 @@
+  case format_specifiert::HASHCODE_UPPER:
+  {
+    format_specifiert fs_lower = fs;
+    fs_lower.conversion = tolower(fs.conversion);
```
actually the upper and lower case length should be the same, so `fs` can be used directly instead of fs_lower

romainbrenguier commented on src/solvers/strings/string_constraint_generator_format.cpp:
```diff
@@ -229,6 +229,7 @@ add_axioms_for_format_specifier(
 /// Deserialize an argument for format from \p string.
 /// \p id should be one of: string_expr, int, char, boolean, float.
 /// The primitive values are expected to all be encoded using 4 characters.
+/// The characters of `string` must be of type `unsignedbv_typet(16)`.
```
can we also add a precondition?

romainbrenguier commented on unit/solvers/strings/string_format_builtin_function/length_for_format_specifier.cpp:
```diff
@@ -0,0 +1,260 @@
+#include <util/string_expr.h>
```
this should come after the comment header

romainbrenguier commented on unit/solvers/strings/string_format_builtin_function/length_for_format_specifier.cpp:
```diff
@@ -0,0 +1,260 @@
+      const array_string_exprt arg = from_char_vector(
+        {0, 0, 0, 12}, char_type, length_type, pointer_type, array_pool);
+      format_specifiert fs{
+        0, "", 0, 0, false, format_specifiert::DECIMAL_INTEGER};
```
:bulb: maybe defining a constructor for `format_specifiert` isn't so useful and writting `format_specifiert fs; fs.conversion = format_specifiert::DECIMAL_INTEGER;` would have been clearer.

romainbrenguier commented on unit/solvers/strings/string_format_builtin_function/length_for_format_specifier.cpp:
```diff
@@ -0,0 +1,260 @@
+      const exprt actual =
+        length_for_format_specifier(fs, arg, length_type, array_pool);
+      const exprt simpl_actual = simplify_expr(actual, ns);
+      THEN("length constraint represents 5")
```
add a comment why it is 5 and not 4 as if the string would be "null"

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b327e6eaa9bcf75877d47c126299e55772f65cb2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/114911867

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 968162be25e84ba5efb1607f1fd54dcf8a16c972).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115080439
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

romainbrenguier commented on src/solvers/strings/string_format_builtin_function.cpp:
```diff
@@ -0,0 +1,482 @@
+  PRECONDITION(serialized.size() == 4);
+
+  // long value, to be used for other formats?
+  for(int i = 0; i < 5; i++)
```
the bound should be 4, and use std::size_t for i

allredj replied:
Oops. Fixed

romainbrenguier commented on src/solvers/strings/string_format_builtin_function.cpp:
```diff
@@ -0,0 +1,482 @@
+  const std::vector<mp_integer> deserialized =
+    make_range(long_as_string).map([&](char c) { return mp_integer{c}; });
+
+  return deserialized;
```
:pick: could return directly the expression above

allredj replied:
done

romainbrenguier commented on src/solvers/strings/string_format_builtin_function.cpp:
```diff
@@ -0,0 +1,482 @@
+          evaluated_char_vector.end(),
+          std::back_inserter(result_vector));
+      }
+      // TODO: handle this case
```
adding the `%%` and `%n` cases should be relatively easy

allredj replied:
Doing now as follow-up PR.

peterschrammel replied:
Are there tickets that track all these todos?

romainbrenguier commented on unit/solvers/strings/string_format_builtin_function/length_for_format_specifier.cpp:
```diff
@@ -0,0 +1,259 @@
+  const auto expr_array =
+    make_range(string_array)
+      .map([&](const mp_integer el) { return from_integer(el, char_type); })
+      .collect<std::vector<exprt>>();
```
:information_source: I think collect is unnecessary if you declare `expr_array` as `std::vector<exprt>` instead of `auto`

allredj replied:
done

romainbrenguier reviewed (approved):
Some nitpicks, but overall it is very good.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: efd585dbdee5ec448a7fd1dd15041829310c5afe).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115085368

peterschrammel commented on src/solvers/strings/format_specifier.cpp:
```diff
@@ -0,0 +1,65 @@
+
+Author: Romain Brenguier, Joel Allred
+
+Date:   June 2019
```
date can be removed

peterschrammel commented on src/solvers/strings/format_specifier.cpp:
```diff
@@ -0,0 +1,65 @@
+
+std::vector<format_elementt> parse_format_string(std::string s)
+{
+  std::string format_specifier =
```
const

peterschrammel commented on src/solvers/strings/format_specifier.h:
```diff
@@ -0,0 +1,161 @@
+    int _width,
+    int _precision,
+    bool _dt,
+    char c)
```
`conversion`

peterschrammel commented on src/solvers/strings/format_specifier.h:
```diff
@@ -0,0 +1,161 @@
+    std::string _flag,
+    int _width,
+    int _precision,
+    bool _dt,
```
What is `dt`?

peterschrammel commented on src/solvers/strings/format_specifier.h:
```diff
@@ -0,0 +1,161 @@
+  char conversion;
+
+  format_specifiert(
+    int _index,
```
underscores are unnecessary

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: cb8f00056438579103085dccd5ace7c706612ff9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115094718

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 49e4ad7660d587d33b316cda4ea45bed821f2dc2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115103659

tautschnig commented on src/solvers/strings/format_specifier.cpp:
```diff
@@ -0,0 +1,65 @@
+  std::string flag = m[2].str();
+  int width = m[3].str().empty() ? -1 : std::stoi(m[3].str());
+  int precision = m[4].str().empty() ? -1 : std::stoi(m[4].str());
+  std::string tT = m[5].str();
```
Wouldn't the index accesses here warrant a `PRECONDITION`?

tautschnig replied:
What does `tT` stand for?

tautschnig commented on src/solvers/strings/format_specifier.cpp:
```diff
@@ -0,0 +1,65 @@
+  std::string tT = m[5].str();
+
+  bool dt = !tT.empty();
+  if(tT == "T")
```
`dt`, `tT` - that's rather magical

tautschnig commented on src/solvers/strings/format_specifier.cpp:
```diff
@@ -0,0 +1,65 @@
+  int index = m[1].str().empty() ? -1 : std::stoi(m[1].str());
+  std::string flag = m[2].str();
+  int width = m[3].str().empty() ? -1 : std::stoi(m[3].str());
+  int precision = m[4].str().empty() ? -1 : std::stoi(m[4].str());
```
Is `int` really the most appropriate type? If anything, an `ssize_t` would seem better suited, but maybe even `optionalt<std::size_t>` to avoid the `-1` hack?

tautschnig commented on src/solvers/strings/format_specifier.cpp:
```diff
@@ -0,0 +1,65 @@
+std::vector<format_elementt> parse_format_string(std::string s)
+{
+  std::string format_specifier =
+    "%(\\d+\\$)?([-#+ 0,(\\<]*)?(\\d+)?(\\.\\d+)?([tT])?([a-zA-Z%])";
```
A comment would be really helpful.

tautschnig commented on src/solvers/strings/format_specifier.cpp:
```diff
@@ -0,0 +1,65 @@
+  std::string format_specifier =
+    "%(\\d+\\$)?([-#+ 0,(\\<]*)?(\\d+)?(\\.\\d+)?([tT])?([a-zA-Z%])";
+  std::regex regex(format_specifier);
+  std::vector<format_elementt> al;
```
What does `al` stand for?

tautschnig commented on src/solvers/strings/format_specifier.cpp:
```diff
@@ -0,0 +1,65 @@
+  return format_specifiert(index, flag, width, precision, dt, conversion);
+}
+
+std::vector<format_elementt> parse_format_string(std::string s)
```
I believe a const reference would be more suitable.

tautschnig commented on src/solvers/strings/format_specifier.cpp:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    al.emplace_back(format_specifier_of_match(match));
+    s = match.suffix();
```
With string views coming, this sort of overwriting of a value isn't a good idea I'd say.

tautschnig reviewed (commented):
Just a couple of observations in a very incomplete review.

peterschrammel commented on src/solvers/strings/string_format_builtin_function.cpp:
```diff
@@ -0,0 +1,489 @@
+  }
+  else
+  {
+    // We only handle 32-bit signed integer type for now.
```
Is there a ticket that describes these limitations?

allredj replied:
For your first comment, this case is now handled (fixed in the fixup commits). For the second one, I don't think this directly translates to a limitation of CBMC. We would only come across it if we venture in supporting String.format for other languages than Java.

allredj replied:
There is no ticket for the other limitations. Do you think we need a ticket for each of them? That would be more than 30 tickets I think.

peterschrammel replied:
It's fine to have a summary ticket or some other form of documentation.

peterschrammel reviewed (approved):
modulo squashing fixups

romainbrenguier committed:
Expose useful functions for format in solver

This will be used by new string_format_builtin_function class.

romainbrenguier committed:
Extract format specifier classes to a new file

Commit:
Fixes in format_specifier

Commit:
add_axioms_for_format now takes array_string_exprts

For the string builtin implementation of string.format, the passed
arguments will already be in array_string_exprt form.

romainbrenguier committed:
Add a class for the string format builtin function

This defines the eval, constraints and length_constraint function, so
that the constraints will only get added when actually needed by the
solver.

romainbrenguier committed:
Use string_format_builtin_function in the solver

Commit:
Remove old add_axioms_for_format

Commit:
Add precondition to format_arg_from_string

Commit:
Use builtin function for String.format

Commit:
Unit tests for length_for_format_specifier

Commit:
Unit tests for length_of_decimal_int

Commit:
Split String.format regression tests

Split by formatter type.

Commit:
Tests for length of String.format of string

Commit:
Tests for length of String.format of decimal integer

Commit:
Tests for length of String.format of hexadecimal integer

Commit:
Tests for length of String.format of boolean

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 18fa75f998ee94ca3fdd43a67d7c9a73a9fc5e4f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115134596

allredj commented:
@tautschnig Thanks for your comments. I'm improving `format_specifier` and its documentation in a follow-up PR.
